korim/src/commonTest/kotlin/com/soywiz/korim/font/TTfTest.kt

```diff
@@ -18,30 +17,75 @@ class TTfTest {
-       fun res(str: String, reader: Boolean = true): Pair<Double, Int> {
-           val rreader = WStringReader(str)
-           val res = font.getGlyphMetrics(100.0, rreader.peek().codePoint, reader = if (reader) rreader else null)
-           return res.xadvance to rreader.position
-       }
-
-       res("s", reader = false).also { (advance, pos) ->
-           assertEquals(42.0, advance, absoluteTolerance = 0.1)
-           assertEquals(0, pos)
-       }
-       res("t", reader = false).also { (advance, pos) ->
-           assertEquals(35.3, advance, absoluteTolerance = 0.1)
-           assertEquals(0, pos)
-       }
-       res("st", reader = true).also { (advance, pos) ->
-           assertEquals(76.4, advance, absoluteTolerance = 0.1)
-           assertEquals(2, pos)
-       }
-   }
+    @Test
+    fun testLigatureAdvancementBug() = suspendTest {
+        val font = resourcesVfs["PlayfairDisplay-BoldItalic.ttf"].readTtfFont()
+        fun res(str: String, reader: Boolean = true): Pair<Double, Int> {
+            val rreader = WStringReader(str)
+            val res = font.getGlyphMetrics(
+                100.0,
+                rreader.peek().codePoint,
+                reader = if (reader) rreader else null
+            )
+            return res.xadvance to rreader.position
+        }
+
+        res("s", reader = false).also { (advance, pos) ->
+            assertEquals(42.0, advance, absoluteTolerance = 0.1)
+            assertEquals(0, pos)
+        }
+        res("t", reader = false).also { (advance, pos) ->
+            assertEquals(35.3, advance, absoluteTolerance = 0.1)
+            assertEquals(0, pos)
+        }
+        res("st", reader = true).also { (advance, pos) ->
+            assertEquals(76.4, advance, absoluteTolerance = 0.1)
+            assertEquals(2, pos)
+        }
+    }
+
+    @Test
+    fun ligaturesEnabledWorks() = suspendTest {
+        val ttfFontWithLigatures =
+            resourcesVfs["font_atkinson/AtkinsonHyperlegible-Bold.ttf"].readTtfFont()
+
+        val reader = WStringReader("1/4")
+
+        val glyphs = mutableListOf<BaseTtfFont.Glyph>()
+
+        while (reader.hasMore) {
+            val c = reader.peek().codePoint
+            val g = ttfFontWithLigatures.getGlyphByReader(reader, c)!!
+            glyphs.add(g)
+        }
+
+        // The string parses to {'¼'}
+        assertEquals(glyphs.size, 1)
+    }
+
+    @Test
+    fun ligaturesDisabledWorks() = suspendTest {
+        val ttfFontWithoutLigatures =
+            resourcesVfs["font_atkinson/AtkinsonHyperlegible-Bold.ttf"].readTtfFont(enableLigatures = false)
+
+        val reader = WStringReader("1/4")
+
+        val glyphs = mutableListOf<BaseTtfFont.Glyph>()
+
+        while (reader.hasMore) {
+            val c = reader.peek().codePoint
+            val g = ttfFontWithoutLigatures.getGlyphByReader(reader, c)!!
+            glyphs.add(g)
+        }
+
+        // The string parses to {'1', '/', '4'}
+        assertEquals(glyphs.size, 3)
```

Comment:
Since this code is duplicated, maybe we can do something to improve it?

What if we add a function inside the TtfFont with a signature like:

```kotin
fun getGlyphsByReader(reader: WStringReader, out: MutableList<Glyph> = arrayListOf()): List<Glyph>
```

Then we do the test like this (and in addition to the number of glyphs we actually ensure that the glyphs emitted are the ones expected):

```kotlin
val ttfFontWithoutLigatures = resourcesVfs["font_atkinson/AtkinsonHyperlegible-Bold.ttf"].readTtfFont(enableLigatures = false)

assertEquals(listOf(...), ttfFontWithoutLigatures.getGlyphsByReader(WStringReader("1/4")).map { it.index })
```

Reply:
Done.